core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+ * under the License.
+ */
+package org.elasticsearch.action.support.replication;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ReplicationResponse;
+import org.elasticsearch.action.UnavailableShardsException;
+import org.elasticsearch.action.WriteConsistencyLevel;
+import org.elasticsearch.action.support.TransportActions;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.transport.TransportResponse;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class ReplicatedOperation<Request extends ReplicationRequest<Request>, ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
+        Response extends ReplicationResponse> {
+    final ESLogger logger;
+    final Request request;
+    final Supplier<ClusterState> clusterStateSupplier;
+    final String opType;
+    final AtomicInteger totalShards = new AtomicInteger();
+    final AtomicInteger pendingShards = new AtomicInteger();
+    final AtomicInteger successfulShards = new AtomicInteger();
+    final boolean executeOnReplicas;
+    final boolean checkWriteConsistency;
+    final Primary<Request, ReplicaRequest, Response> primary;
+    final ReplicasProxy<ReplicaRequest> replicasProxy;
+    final AtomicBoolean finished = new AtomicBoolean();
+    final ActionListener<Response> finalResponseListener;
+
+    volatile Response finalResponse = null;
+
+    private final List<ReplicationResponse.ShardInfo.Failure> shardReplicaFailures = Collections.synchronizedList(new ArrayList<>());
+
+
+    public ReplicatedOperation(Request request, Primary<Request, ReplicaRequest, Response> primary,
+                               ActionListener<Response> listener,
+                               boolean executeOnReplicas, boolean checkWriteConsistency,
+                               ReplicasProxy<ReplicaRequest> replicasProxy,
+                               Supplier<ClusterState> clusterStateSupplier, ESLogger logger, String opType) {
+        this.checkWriteConsistency = checkWriteConsistency;
+        Objects.requireNonNull(request.shardId(), "missing shardId");
```

Comment:
At a second thought, I removed all usages of request.shardId(), so we don't care any more.